File: src/main/java/io/dddrive/doc/model/enums/CodeCaseDef.java

package io.dddrive.doc.model.enums;

import lombok.Data;
import lombok.experimental.SuperBuilder;

import java.util.ArrayList;
import java.util.List;

import com.fasterxml.jackson.annotation.JsonIgnore;
import io.dddrive.enums.model.base.EnumeratedBase;

@Data
@SuperBuilder
public class CodeCaseDef extends EnumeratedBase {

	public final List<CodeCaseStage> caseStages = new ArrayList<>();

	public CodeCaseDef(CodeCaseDefEnum enumeration, String id, String name) {
		super(enumeration, id, name);
	}

	public void addCaseStage(CodeCaseStage stage) {
		this.caseStages.add(stage);
	}

	@JsonIgnore
	public List<CodeCaseStage> getCaseStages() {
		return List.copyOf(this.caseStages);
	}

	@Override
	public boolean equals(Object other) {
		return super.equals(other);
	}

	@Override
	public int hashCode() {
		return super.hashCode();
	}

}

File: fm-dddrive/src/main/java/dddrive/app/doc/model/enums/CodeCaseDef.kt
package dddrive.app.doc.model.enums

import dddrive.ddd.model.Enumerated
import dddrive.ddd.model.enums.CodeAggregateType
import dddrive.ddd.model.enums.CodeAggregateTypeEnum.Companion.getAggregateType

class CodeCaseDef(
	override val id: String,
	override val defaultName: String,
	private val docTypeId: String,
) : Enumerated {

	val docType: CodeAggregateType get() = getAggregateType(docTypeId)

	private val _caseStages: MutableList<CodeCaseStage> = mutableListOf()

	override val enumeration: CodeCaseDefEnum
		get() = CodeCaseDefEnum.instance

	fun addCaseStage(stage: CodeCaseStage) = _caseStages.add(stage)

	val caseStages: List<CodeCaseStage> get() = _caseStages.toList()

}
